File: 1163-last-substring-in-lexicographical-order/1163-last-substring-in-lexicographical-order.py
class Solution:
    def lastSubstring(self, s: str) -> str:
        l, r, offset = 0, 1, 0
        while r + offset < len(s):
            if s[l + offset] == s[r + offset]:
                offset += 1
            else:
                if s[l + offset] < s[r + offset]:
                    l += offset + 1
                else:
                    r += offset + 1
                if l == r:
                    r += 1
                offset = 0
        return s[l:]

File: 1163-last-substring-in-lexicographical-order/1163-last-substring-in-lexicographical-order.js
/**
 * @param {string} s
 * @return {string}
 */
var lastSubstring = function(s) {
  let [l, r, offset] = [0, 1, 0];
  
  while (r + offset < s.length) {
    const [charL, charR] = [s.charCodeAt(l + offset), s.charCodeAt(r + offset)];
    if (charL === charR) ++offset;
    else {
      charL < charR ? (l += offset + 1) : (r += offset + 1);
      
      if (l === r) ++r;
      offset = 0;
    }
  }
  
  return s.slice(l);
};